lib/spack/spack/build_environment.py

```diff
@@ -752,19 +752,13 @@ def setup_package(pkg, dirty, context: Context = Context.BUILD):
     target = platform.target(pkg.spec.architecture.target)
     platform.setup_platform_environment(pkg, env_mods)
 
-    if context == Context.BUILD:
-        tty.debug("setup_package: setup build environment for root")
-        builder = spack.builder.create(pkg)
-        builder.setup_build_environment(env_mods)
```

Comment:
I'm still trying to convince myself that removing this is a good idea or doesn't do harm to post-install testing.

Reply:
It should in principle be the same, `setup_build_environment` is called when `context == BUILD` in `get_env_modifications` now, and that function is called a few lines up on line 747.

Reply:
Speaking of post-install tests, I find it rather confusing... `spack install --test` is meant for build time tests, correct? And `spack test run` is meant for install time tests?

But Python packages by default do import tests post-install, which is an install time test. And that certainly needs the runtime environment variables to be set up, cause pure run deps are missing from PATH during the build.

Reply:
> Speaking of post-install tests, I find it rather confusing... `spack install --test` is meant for build time tests, correct? 

It's meant for any post-phase tests packages want to add, where the phases of interest are `build` and `install`).  How to set this up with custom tests and which build systems automatically do this is discussed in https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-installation-phase-tests with the table showing build systems that automatically invoke post-phase checks (when present).

There was a time, pre-multiple build systems, where packages could simply specify the `build_time_test_callbacks` and or `install_time_test_callbacks` properties to either increase or override/disable build system checks.  Build system packages are still specifying these properties though they are no longer working as expected (without additional work).

>And `spack test run` is meant for install time tests?

This is intended to run `def test*` methods post-install originally separate from post-`install` phase checks but we saw no reason to preclude them if/when it makes sense.  These are discussed in the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#stand-alone-tests.

> 
> But Python packages by default do import tests post-install, which is an install time test. And that certainly needs the runtime environment variables to be set up, cause pure run deps are missing from PATH during the build.

That is correct.  This was a request made by @adamjstewart not long after the original introduction of stand-alone tests.

Reply:
This also affects the `SIPPackage`.


I'm still working on a more general solution to provide the behavior we had initially in https://github.com/spack/spack/pull/40568.  Manual testing of the first commit -- python only -- seemed to work, at least at first, then I started working on a prototype of a more general solution.

Reply:
> > And `spack test run` is meant for install time tests?
> 
> This is intended to run `def test*` methods post-install originally separate from post-`install` phase checks but we saw no reason to preclude them if/when it makes sense.

The issue is that the build environment can be different from the run environment. They can overlap, but the build environment may include things not in the run environment and vice versa. So, if you insist on running those tests post install, we'd have to setup the run environment. That said, there was talk about relaxing the requirement that run deps need to be installed before the dependent, which would A. solve an issue in the python ecosystem with circular deps among pure run dependencies (not that the solver supports that now...) and B. lead to more parallelism as pure run deps can be installed concurrently with their dependent in another process.

> This also affects the `SIPPackage`.

I don't understand whether this PR is breaking anything

Reply:
> > > And `spack test run` is meant for install time tests?
> > 
> > 
> > This is intended to run `def test*` methods post-install originally separate from post-`install` phase checks but we saw no reason to preclude them if/when it makes sense.
> 
> The issue is that the build environment can be different from the run environment. They can overlap, but the build environment may include things not in the run environment and vice versa. So, if you insist on running those tests post install, we'd have to setup the run environment.

Yes, we do need the run environment after the post-install phase.  It was being set up for running the post-install phase tests previously.

> That said, there was talk about relaxing the requirement that run deps need to be installed before the dependent, which would A. solve an issue in the python ecosystem with circular deps among pure run dependencies (not that the solver supports that now...) and B. lead to more parallelism as pure run deps can be installed concurrently with their dependent in another process.
> 
> > This also affects the `SIPPackage`.
> 
> I don't understand whether this PR is breaking anything

I didn't notice a difference in my initial testing BUT let me double check with the simple Python "fix".

Reply:
FYI.  My "quick" fix still doesn't resolve the adapter issue for `spack install --test` so that isn't the right case for this issue.